### ISSUE TITLE
Capabilities & Security (custom cap + filters)

### ISSUE THREAD

**Issue by SitesByYogi** [SitesByYogi/nfinite-wp-audit#8]
## Scope
- Create custom capability `manage_nfinite_audit` and assign to Administrator on activation.
- Gate all menus/pages/AJAX via helper `Nfinite_Audit_Admin::cap()` with `apply_filters('nfinite_audit_capability', ...)`.
- Keep nonce patterns consistent across AJAX/forms.

## Stub (drop-in)
```php
// main plugin file
register_activation_hook(__FILE__, function(){
    $role = get_role('administrator');
    if ( $role && ! $role->has_cap('manage_nfinite_audit') ) {
        $role->add_cap('manage_nfinite_audit');
    }
});

// In Admin class
public static function cap() {
    return apply_filters('nfinite_audit_capability', 'manage_nfinite_audit');
}
// Use self::cap() for add_menu_page, add_submenu_page, AJAX callbacks, etc.

**Comment by SitesByYogi** [comment]
### Acceptance Criteria

- Non-admins without the cap cannot access pages/AJAX.

- Sites can re-map capability via filter.

### Test Steps

- Create Editor; confirm no access.

- Grant the cap; confirm access works.

**Comment by SitesByYogi** [comment]
### 1) Main plugin file (bootstrap) — capabilities + mapping

Add this near the top (after your constants):


```
// === Capabilities ===
if ( ! defined('NFINITE_AUDIT_CAP') ) {
    define('NFINITE_AUDIT_CAP', 'manage_nfinite_audit');
}

/**
 * On activation: grant the custom cap to Administrators.
 * Handles both single-site and network activation.
 */
register_activation_hook(__FILE__, function($network_wide = false){
    if ( is_multisite() && $network_wide ) {
        // Grant on each site in the network
        $sites = get_sites(['fields' => 'ids']);
        foreach ( $sites as $blog_id ) {
            switch_to_blog($blog_id);
            $role = get_role('administrator');
            if ( $role && ! $role->has_cap(NFINITE_AUDIT_CAP) ) {
                $role->add_cap(NFINITE_AUDIT_CAP);
            }
            restore_current_blog();
        }
    } else {
        $role = get_role('administrator');
        if ( $role && ! $role->has_cap(NFINITE_AUDIT_CAP) ) {
            $role->add_cap(NFINITE_AUDIT_CAP);
        }
    }
});

/**
 * Optional: map our custom cap to a core cap so admins keep access
 * even if the role edit/remove happens. This is filterable so sites
 * can harden or loosen requirements.
 */
add_filter('map_meta_cap', function($caps, $cap, $user_id, $args){
    if ( $cap !== NFINITE_AUDIT_CAP ) { 
        return $caps; 
    }

    // Default requirement: manage_options (Administrator by default).
    // Site owners can adjust via filter below.
    $required = apply_filters('nfinite_audit_required_caps', ['manage_options']);

    // Normalize to array
    $required = is_array($required) ? $required : [$required];

    return $required;
}, 10, 4);

/**
 * Convenience filter so other plugins/themes can remap the single
 * capability name we check throughout the plugin (menus/AJAX/forms).
 * Example override in a must-use plugin:
 *    add_filter('nfinite_audit_capability', fn() => 'manage_options');
 */
add_filter('nfinite_audit_capability', function($cap){
    return NFINITE_AUDIT_CAP;
});
```

**Notes**
• We grant the custom cap to Administrators at activation.
• We map `manage_nfinite_audit` to `manage_options` by default via `map_meta_cap` so Admins never get locked out even if the role cap was removed later.
• Site owners can harden requirements with:

`add_filter('nfinite_audit_required_caps', fn() => ['manage_options','install_plugins']);
`
• Or remap which single cap the plugin checks with:

`add_filter('nfinite_audit_capability', fn() => 'manage_options');
`
(You typically don’t remove caps on deactivation so audit access settings survive upgrades; if you prefer to clean up, you can add a `register_deactivation_hook` to remove it.)

### 2) Admin class helpers — single source of truth

In admin/class-admin.php (or wherever Nfinite_Audit_Admin lives), add:
class Nfinite_Audit_Admin {

 
```
// ... existing properties/methods ...

    /**
     * Return the capability string the plugin uses for gating.
     * Filterable via 'nfinite_audit_capability'.
     */
    public static function cap() {
        $cap = apply_filters('nfinite_audit_capability', defined('NFINITE_AUDIT_CAP') ? NFINITE_AUDIT_CAP : 'manage_options');
        return $cap ?: 'manage_options';
    }

    /**
     * Convenience wrapper for current_user_can on our cap.
     */
    public static function can() {
        return current_user_can( self::cap() );
    }

    /**
     * Guard utility for admin pages and handlers.
     * Dies with 403 if capability missing.
     */
    public static function require_cap() {
        if ( ! self::can() ) {
            wp_die( esc_html__('You do not have permission to access this page.', 'nfinite-site-audit'), 403 );
        }
    }

    /**
     * Standardized nonce action key for this plugin.
     * Use per-action suffixes like 'run_audit', 'export_json', etc.
     */
    public static function nonce_action( $suffix ) {
        return "nfinite_{$suffix}";
    }

    /**
     * Render a nonce field for forms (admin screens).
     */
    public static function nonce_field( $suffix ) {
        wp_nonce_field( self::nonce_action($suffix), "_wpnonce_{$suffix}" );
    }

    /**
     * Verify nonce for admin_post/wp_ajax requests.
     * Dies with 400 on failure.
     */
    public static function verify_nonce_or_die( $suffix ) {
        $key = "_wpnonce_{$suffix}";
        $action = self::nonce_action($suffix);

        // AJAX uses check_ajax_referer; admin_post can also use it — both return/exit on failure.
        if ( defined('DOING_AJAX') && DOING_AJAX ) {
            check_ajax_referer( $action, $key );
        } else {
            if ( ! isset($_REQUEST[$key]) || ! wp_verify_nonce( sanitize_text_field(wp_unslash($_REQUEST[$key])), $action ) ) {
                wp_die( esc_html__('Invalid request. Please try again.', 'nfinite-site-audit'), 400 );
            }
        }
    }

    // Example: hook registration (menus/AJAX) would call self::cap() below...
}
```
### 3) Gate all menus/pages with the helper cap

Where you register menus:

```
add_action('admin_menu', function(){
    $cap = Nfinite_Audit_Admin::cap();

    add_menu_page(
        __('Nfinite Audit','nfinite-site-audit'),
        __('Nfinite Audit','nfinite-site-audit'),
        $cap,
        'nfinite-audit',
        'nfinite_audit_render_dashboard',
        'dashicons-performance',
        59
    );

    add_submenu_page(
        'nfinite-audit',
        __('SEO Basics','nfinite-site-audit'),
        __('SEO Basics','nfinite-site-audit'),
        $cap,
        'nfinite-audit-seo',
        'nfinite_audit_render_seo'
    );

    // ...other pages, always using $cap...
});
```
Inside each render callback, hard-guard as well:

```
function nfinite_audit_render_dashboard(){
    Nfinite_Audit_Admin::require_cap();

    echo '<div class="wrap">';
    echo '<h1>' . esc_html__('Nfinite Audit', 'nfinite-site-audit') . '</h1>';

    // Example form with standardized nonce:
    echo '<form method="post" action="' . esc_url( admin_url('admin-post.php') ) . '">';
    echo '<input type="hidden" name="action" value="nfinite_run_audit_now" />';
    Nfinite_Audit_Admin::nonce_field('run_audit');
    submit_button( __('Run Audit Now','nfinite-site-audit') );
    echo '</form>';

    echo '</div>';
}
```

### 4) Protect admin_post and AJAX endpoints

```
admin_post (you already use `admin_post_nfinite_run_audit_now`):

```
```
add_action('admin_post_nfinite_run_audit_now', function(){
    Nfinite_Audit_Admin::require_cap();
    Nfinite_Audit_Admin::verify_nonce_or_die('run_audit');

    // ... perform the audit ...
    // On success:
    wp_safe_redirect( add_query_arg('nfinite_audit', 'ran', wp_get_referer() ?: admin_url('admin.php?page=nfinite-audit') ) );
    exit;
});
```
AJAX (example: `wp_ajax_nfinite_export_json`):

```
add_action('wp_ajax_nfinite_export_json', function(){
    Nfinite_Audit_Admin::require_cap();
    Nfinite_Audit_Admin::verify_nonce_or_die('export_json');

    // ... build $payload ...
    wp_send_json_success([
        'filename' => 'nfinite-audit.json',
        'data'     => $payload,
    ]);
});
```
When enqueueing buttons/links that hit AJAX, include the nonce:
```
wp_localize_script('nfinite-admin', 'NFINITE_AUDIT', [
    'ajax_url' => admin_url('admin-ajax.php'),
    'nonces'   => [
        'export_json' => wp_create_nonce( Nfinite_Audit_Admin::nonce_action('export_json') ),
    ],
    'cap'      => Nfinite_Audit_Admin::cap(),
]);
```
**JS usage example:**

```
jQuery(function($){
  $('#nfinite-export').on('click', function(e){
    e.preventDefault();
    $.post( NFINITE_AUDIT.ajax_url, {
      action: 'nfinite_export_json',
      _wpnonce_export_json: NFINITE_AUDIT.nonces.export_json
    }, function(res){
      if (res.success) {
        // trigger download...
      } else {
        alert('Export failed.');
      }
    });
  });
});
```

### 5) Consistent nonce pattern

Action key: `nfinite_{verb}` (e.g., `nfinite_run_audit`, `nfinite_export_json`).

Field name: "`_wpnonce_{verb}`" (e.g., `_wpnonce_run_audit`).

Create: `Nfinite_Audit_Admin::nonce_field('{verb}'`)

Verify: `Nfinite_Audit_Admin::verify_nonce_or_die('{verb}'`)

This keeps forms and AJAX uniform and easy to audit.

### 6) Filter examples for site owners (docs snippet)

These show how the issue’s “Sites can re-map capability via filter” acceptance criterion is met:

```
// 1) Remap the single capability the plugin checks (menus/AJAX).
add_filter('nfinite_audit_capability', function($default){
    // e.g., only Super Admins on multisite:
    return is_multisite() ? 'manage_network' : $default;
});

// 2) Harden the underlying meta-cap mapping.
add_filter('nfinite_audit_required_caps', function($caps){
    // Require both manage_options and install_plugins
    return ['manage_options','install_plugins'];
});
```
### 7) Acceptance criteria checklist (for the PR)

 Non-admins without the cap cannot access pages/AJAX.
Verified via `current_user_can( self::cap() )` guard and `map_meta_cap` fallback.

 Sites can re-map capability via filter.
`nfinite_audit_capability` + `nfinite_audit_required_caps` provided.

 Nonce patterns consistent across AJAX/forms.
`nonce_action('suffix')`, `nonce_field('suffix')`, and `verify_nonce_or_die('suffix')`.

### 8) Test steps (copy into the issue)

Create Editor user on the site.

As Editor, navigate to **Nfinite Audit menu** → expect permission denied.

Hit an AJAX endpoint (e.g., Export JSON) → expect 403.

Grant the cap to the Editor:

Temporarily add in a mu-plugin or run via WP-CLI:

```
$role = get_role('editor');
$role && $role->add_cap('manage_nfinite_audit');
```

Filter remap check (optional):

**Add:**

`add_filter('nfinite_audit_capability', fn() => 'manage_options');
`

Remove custom cap from Editor; confirm access now depends on manage_options.
